File: psort.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"syscall"
	"unsafe"
)

var reWhitespace = regexp.MustCompile(`\s+`)

type winsize struct{ Rows, Cols, X, Y uint16 }

func termWidth() int {
	if s := os.Getenv("COLUMNS"); s != "" {
		if n, err := strconv.Atoi(s); err == nil && n > 0 {
			return n
		}
	}
	var ws winsize
	r, _, _ := syscall.Syscall(syscall.SYS_IOCTL, 1, 0x5413, uintptr(unsafe.Pointer(&ws)))
	if r == 0 && ws.Cols > 0 {
		return int(ws.Cols)
	}
	return 80
}

func makeBar(p float64, barWidth int) string {
	filled := int(p / 100.0 * float64(barWidth))
	if filled > barWidth {
		filled = barWidth
	}
	return "[" + strings.Repeat("#", filled) + strings.Repeat(" ", barWidth-filled) + "]"
}

func main() {
	csv := flag.Bool("csv", false, "Do output as csv rather than formatted for a human to read")
	flag.BoolVar(csv, "c", false, "")
	csvall := flag.Bool("csvall", false, "Do output as csv replacing any/all whitespaces with commas")
	flag.BoolVar(csvall, "a", false, "")
	decimal := flag.Int("decimal", 4, "Specify how many decimal places to compute (default 4)")
	flag.IntVar(decimal, "d", 4, "")
	numeric := flag.Bool("numeric", false, "Sort output by numeric value of data instead of by counts")
	mincount := flag.Int("min", 0, "Minimum count; entries below this are excluded")
	flag.IntVar(mincount, "m", 0, "")
	maxcount := flag.Int("max", 0, "Maximum count; entries above this are excluded")
	flag.IntVar(maxcount, "x", 0, "")
	fullperc := flag.Bool("fullperc", false, "With min/max, compute % based on full count rather than filtered count")
	flag.BoolVar(fullperc, "f", false, "")
	ignoreExcluded := flag.Bool("ignore", false, "With min/max, suppress the [excluded] summary line")
	flag.BoolVar(ignoreExcluded, "i", false, "")
	verbose := flag.Bool("verbose", false, "With min/max, print key/count totals before and after filtering")
	flag.BoolVar(verbose, "v", false, "")
	reverse := flag.Bool("reverse", false, "Reverse the sort order (highest count/value first)")
	flag.BoolVar(reverse, "r", false, "")
	histogram := flag.Bool("histogram", false, "Show a bar histogram scaled to terminal width")
	flag.BoolVar(histogram, "H", false, "")
	flag.Usage = usage
	flag.Parse()

	if *fullperc && *mincount == 0 && *maxcount == 0 {
		fmt.Fprintln(os.Stderr, "ERROR: cannot use fullperc without either min or max")
		usage()
		os.Exit(1)
	}

	counts := make(map[string]int)
	var tot int

	readLines := func(f *os.File) {
		scanner := bufio.NewScanner(f)
		for scanner.Scan() {
			line := scanner.Text()
			counts[line]++
			tot++
		}
	}

	if flag.NArg() > 0 {
		for _, path := range flag.Args() {
			f, err := os.Open(path)
			if err != nil {
				fmt.Fprintln(os.Stderr, err)
				continue
			}
			readLines(f)
			f.Close()
		}
	} else {
		readLines(os.Stdin)
	}

	if tot == 0 {
		return
	}

	if *mincount > 0 || *maxcount > 0 {
		if *verbose {
			tc := 0
			for _, v := range counts {
				tc += v
			}
			fmt.Fprintf(os.Stderr, "# keys/counts before: %d / %d\n", len(counts), tc)
		}

		excluded := 0
		for key, count := range counts {
			drop := (*mincount > 0 && count < *mincount) || (*maxcount > 0 && count > *maxcount)
			if drop {
				excluded += count
				if !*fullperc {
					tot -= count
				}
				delete(counts, key)
			}
		}

		if excluded > 0 && !*ignoreExcluded {
			counts["[excluded]"] += excluded
		}

		if *verbose {
			tc := 0
			for _, v := range counts {
				tc += v
			}
			fmt.Fprintf(os.Stderr, "# keys/counts after: %d / %d\n", len(counts), tc)
		}
	}

	if tot == 0 {
		return
	}

	keys := make([]string, 0, len(counts))
	for k := range counts {
		keys = append(keys, k)
	}

	if *numeric {
		sort.Slice(keys, func(i, j int) bool {
			vi, _ := strconv.ParseFloat(keys[i], 64)
			vj, _ := strconv.ParseFloat(keys[j], 64)
			if *reverse {
				return vi > vj
			}
			return vi < vj
		})
	} else {
		sort.Slice(keys, func(i, j int) bool {
			if *reverse {
				return counts[keys[i]] > counts[keys[j]]
			}
			return counts[keys[i]] < counts[keys[j]]
		})
	}

	lenTot := len(strconv.Itoa(tot))

	barWidth := 0
	if *histogram && !*csv && !*csvall {
		maxKey := 0
		for _, k := range keys {
			if len(k) > maxKey {
				maxKey = len(k)
			}
		}
		barWidth = termWidth() - *decimal - lenTot - maxKey - 11
		if barWidth < 10 {
			barWidth = 10
		}
	}

	for _, key := range keys {
		pout(key, counts[key], tot, lenTot, *decimal, *csv, *csvall, *histogram, barWidth)
	}

	if *histogram && !*csv && !*csvall {
		fmt.Printf("%*s  %*d\n", barWidth+3+*decimal+4, "Total:", lenTot, tot)
	} else if *csv || *csvall {
		fmt.Printf("%8s,%*d\n", "Total:", lenTot, tot)
	} else {
		fmt.Printf("%8s  %*d\n", "Total:", lenTot, tot)
	}
}

func pout(key string, count, tot, lenTot, decimal int, csv, csvall, histogram bool, barWidth int) {
	p := (float64(count) / float64(tot)) * 100.0

	outKey := key
	if csvall {
		outKey = reWhitespace.ReplaceAllString(key, ",")
	}

	width := 2
	if p < 10 {
		width = 1
	}
	if csv || csvall {
		fmt.Printf("%*.*f%%,%d,%s\n", width, decimal, p, count, outKey)
	} else if histogram {
		var percStr string
		if p < 10 {
			percStr = " " + fmt.Sprintf("%.*f%%", decimal, p)
		} else {
			percStr = fmt.Sprintf("%.*f%%", decimal, p)
		}
		fmt.Printf("%s %s  %*d %s\n", makeBar(p, barWidth), percStr, lenTot, count, outKey)
	} else {
		if p < 10 {
			fmt.Printf(" %*.*f%%  %*d %s\n", width, decimal, p, lenTot, count, outKey)
		} else {
			fmt.Printf("%*.*f%%  %*d %s\n", width, decimal, p, lenTot, count, outKey)
		}
	}
}

func usage() {
	fmt.Fprintf(os.Stderr, `Usage: %s [options] [file ...]
    --csv, -c    : Do output as csv rather than formatted for a human to read
    --csvall, -a : Do output as csv replacing any/all whitespaces with commas
    --decimal, -d: Specify how many decimal places to compute (default 4)
    --numeric    : If the data being fed to psort is numeric this flag will sort the
                 :  output by that data instead of by counts
    --min, -m    : Give a minimum count below which we should throw out those, to
                    allow trimming smaller values (compute %% based upon what's left)
    --max, -x    : Give a maximum count above which we should throw out those, to
                    allow trimming larger values  (compute %% based upon what's left)
    --fullperc, -f: Only with a min/max, still output the %% based upon the full count
                    vs only what matched the filter
    --ignore, -i : When using min/max, don't include the [excluded] line
                    that shows us the full 100%%
    --verbose, -v: When using min/max, print key/count totals before and after filtering
    --reverse,-r : Reverse the sort order (highest count/value first)
    --histogram,-H: Show a bar histogram scaled to terminal width
    --help, -h   : Output this help info

Synopsis: Takes input and essentially does the equivalent of `+"`"+`sort | uniq -c | sort -n`+"`"+`
with added percentages and total info, default output is nicely formatted for humans to read.
`, os.Args[0])
}

File: psort.py
#!/usr/bin/env python3
"""psort — sort | uniq -c | sort -n with percentages."""

import argparse
import re
import sys


def pout(key, count, tot, len_tot, decimal, csv_mode, csvall):
    p = (count / tot) * 100.0
    out_key = re.sub(r'\s+', ',', key) if csvall else key
    p_str = f"{p:.{decimal}f}"
    if csv_mode or csvall:
        print(f"{p_str}%,{count},{out_key}")
    else:
        if p < 10:
            print(f" {p_str}%  {count:{len_tot}d} {out_key}")
        else:
            print(f"{p_str}%  {count:{len_tot}d} {out_key}")


def main():
    parser = argparse.ArgumentParser(
        prog='psort',
        add_help=False,
        formatter_class=argparse.RawDescriptionHelpFormatter,
        usage='%(prog)s [options] [file ...]',
        description=(
            "Synopsis: Takes input and essentially does the equivalent of\n"
            "`sort | uniq -c | sort -n` with added percentages and total info,\n"
            "default output is nicely formatted for humans to read."
        ),
    )
    parser.add_argument('-c', '--csv',      action='store_true', help='Do output as csv rather than formatted for a human to read')
    parser.add_argument('-a', '--csvall',   action='store_true', help='Do output as csv replacing any/all whitespaces with commas')
    parser.add_argument('-d', '--decimal',  type=int, default=4, metavar='N', help='Specify how many decimal places to compute (default 4)')
    parser.add_argument(      '--numeric',  action='store_true', help='Sort output by numeric value of data instead of by counts')
    parser.add_argument('-m', '--min',      type=int, default=0, metavar='N', help='Minimum count; entries below this are excluded')
    parser.add_argument('-x', '--max',      type=int, default=0, metavar='N', help='Maximum count; entries above this are excluded')
    parser.add_argument('-f', '--fullperc', action='store_true', help='With min/max, compute %% based on full count rather than filtered count')
    parser.add_argument('-i', '--ignore',   action='store_true', help='With min/max, suppress the [excluded] summary line')
    parser.add_argument('-v', '--verbose',  action='store_true', help='With min/max, print key/count totals before and after filtering')
    parser.add_argument('-h', '--help',     action='help',       help='Output this help info')
    parser.add_argument('files', nargs='*', metavar='file')

    args = parser.parse_args()

    if args.fullperc and args.min == 0 and args.max == 0:
        print("ERROR: cannot use fullperc without either min or max", file=sys.stderr)
        parser.print_usage(sys.stderr)
        sys.exit(1)

    counts = {}
    tot = 0

    def read_lines(f):
        nonlocal tot
        for line in f:
            line = line.rstrip('\n').rstrip('\r')
            counts[line] = counts.get(line, 0) + 1
            tot += 1

    if args.files:
        for path in args.files:
            try:
                with open(path) as f:
                    read_lines(f)
            except OSError as e:
                print(e, file=sys.stderr)
    else:
        read_lines(sys.stdin)

    if tot == 0:
        return

    if args.min > 0 or args.max > 0:
        if args.verbose:
            print(f"# keys/counts before: {len(counts)} / {sum(counts.values())}", file=sys.stderr)

        excluded = 0
        for key in list(counts):
            count = counts[key]
            if (args.min > 0 and count < args.min) or (args.max > 0 and count > args.max):
                excluded += count
                if not args.fullperc:
                    tot -= count
                del counts[key]

        if excluded > 0 and not args.ignore:
            counts['[excluded]'] = counts.get('[excluded]', 0) + excluded

        if args.verbose:
            print(f"# keys/counts after: {len(counts)} / {sum(counts.values())}", file=sys.stderr)

    if tot == 0:
        return

    if args.numeric:
        def numeric_key(k):
            try:
                return float(k)
            except ValueError:
                return 0.0
        keys = sorted(counts, key=numeric_key)
    else:
        keys = sorted(counts, key=lambda k: counts[k])

    len_tot = len(str(tot))

    for key in keys:
        pout(key, counts[key], tot, len_tot, args.decimal, args.csv, args.csvall)

    if args.csv or args.csvall:
        print(f"{'Total:':>8},{tot:{len_tot}d}")
    else:
        print(f"{'Total:':>8}  {tot:{len_tot}d}")


if __name__ == '__main__':
    main()
